### PR TITLE
wip: support jax rs client instrumentation

### PR DESCRIPTION
**Pull request by kananindzya**
### Checklist

- [ ] Implement code : sync requests
- [ ] Add tests
- [ ] Update documentation
- [ ] Update [CHANGELOG.md](CHANGELOG.md)
- [ ] Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)

### CODE REVIEW

**Comment by felixbarny** [comment]
As explained in https://github.com/elastic/apm-agent-java/issues/713, we don't want to instrument at the JAX-RS client API level but rather want to instrument the underlying HttpClient. In that case it's HttpUrlConnection, which we already support.

File: apm-agent-plugins/apm-jaxrs-httpclient-plugin/src/test/java/co/elastic/apm/agent/jaxrs/httpclient/JaxRsHttpClientInstrumentationTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    @Before
+    public void setUp() {
+        this.client = ClientBuilder.newClient();
+    }
+
+    @Override
+    protected void performGet(String path) throws Exception {
+        Response response = client
+            .target(path)
+            .request(MediaType.APPLICATION_JSON)
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-jaxrs-httpclient-plugin/src/test/java/co/elastic/apm/agent/jaxrs/httpclient/JaxRsHttpClientInstrumentationTest.java]:
```suggestion
            .request(MediaType.APPLICATION_JSON)
            .async()
            .get()
```

Does this pass as well?

Comment (kananindzya) [reply apm-agent-plugins/apm-jaxrs-httpclient-plugin/src/test/java/co/elastic/apm/agent/jaxrs/httpclient/JaxRsHttpClientInstrumentationTest.java]: No, it doesn't.
```
main] INFO wiremock.org.eclipse.jetty.server.handler.ContextHandler - Started w.o.e.j.s.ServletContextHandler@75aea2ba{/__admin,null,AVAILABLE}
[main] INFO wiremock.org.eclipse.jetty.server.handler.ContextHandler - Started w.o.e.j.s.ServletContextHandler@5ff6dd3c{/,null,AVAILABLE}
[main] INFO wiremock.org.eclipse.jetty.server.NetworkTrafficServerConnector - Started NetworkTrafficServerConnector@5fcd1f0a{HTTP/1.1}{0.0.0.0:34203}
[main] INFO wiremock.org.eclipse.jetty.server.Server - Started @6918ms
[main] DEBUG co.elastic.apm.agent.impl.transaction.AbstractSpan - increment references to '' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4) (1)
[main] DEBUG co.elastic.apm.agent.impl.ElasticApmTracer - startTransaction '' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4) {
[main] DEBUG co.elastic.apm.agent.impl.transaction.AbstractSpan - increment references to '' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4) (2)
[main] DEBUG co.elastic.apm.agent.impl.ElasticApmTracer - Activating '' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4) on thread 1
[qtp137685382-62] INFO / - RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.StubRequestHandler. Normalized mapped under returned 'null'
[main] DEBUG co.elastic.apm.agent.impl.ElasticApmTracer - Deactivating '' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4) on thread 1
[main] DEBUG co.elastic.apm.agent.impl.transaction.AbstractSpan - decrement references to '' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4) (1)
[main] DEBUG co.elastic.apm.agent.impl.ElasticApmTracer - } endTransaction 'unnamed' 00-4b9e8b4c67e4bf89081eddf7b1db6e34-9708b53297d4bf83-01 (536b71b4)
[main] INFO wiremock.org.eclipse.jetty.server.NetworkTrafficServerConnector - Stopped NetworkTrafficServerConnector@5fcd1f0a{HTTP/1.1}{0.0.0.0:0}
[main] INFO wiremock.org.eclipse.jetty.server.handler.ContextHandler - Stopped w.o.e.j.s.ServletContextHandler@5ff6dd3c{/,null,UNAVAILABLE}
[main] INFO wiremock.org.eclipse.jetty.server.handler.ContextHandler - Stopped w.o.e.j.s.ServletContextHandler@75aea2ba{/__admin,null,UNAVAILABLE}
[main] WARN wiremock.org.eclipse.jetty.util.thread.QueuedThreadPool - qtp137685382{STOPPING,8<=8<=10,i=7,q=6} Couldn't stop Thread[qtp137685382-64,5,main]
[main] WARN wiremock.org.eclipse.jetty.util.thread.QueuedThreadPool - qtp137685382{STOPPING,8<=8<=10,i=3,q=6} Couldn't stop Thread[qtp137685382-59,5,]
[main] WARN wiremock.org.eclipse.jetty.util.thread.QueuedThreadPool - qtp137685382{STOPPING,8<=8<=10,i=0,q=5} Couldn't stop Thread[qtp137685382-68,5,]

java.lang.IndexOutOfBoundsException: Index 0 out-of-bounds for length 0

	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
```

Comment (eyalkoren) [reply apm-agent-plugins/apm-jaxrs-httpclient-plugin/src/test/java/co/elastic/apm/agent/jaxrs/httpclient/JaxRsHttpClientInstrumentationTest.java]: So, I guess this reproduces the problem originally reported in https://discuss.elastic.co/t/jaxrs-async-distrubuted-tracing-not-works/189531, right?

**Comment by eyalkoren** [comment]
@kananindzya since the JAX-RS instrumentation relies on low level stuff like `HttpUrlConnection` and `Executor`s, we cannot test it through the regular unit tests, where agent classes are loaded from the system class loader. We need to test it with real agent installation, where agent classes are loaded by Bootstrap class loader, which is used to load the JVM classes.

In addition, we decided not to instrument the JAX-RS APIs, as explained above, so I want to close this PR and we will start over once we decide to address #713 .

Is that OK?

**Comment by kananindzya** [comment]
> @kananindzya since the JAX-RS instrumentation relies on low level stuff like `HttpUrlConnection` and `Executor`s, we cannot test it through the regular unit tests, where agent classes are loaded from the system class loader. We need to test it with real agent installation, where agent classes are loaded by Bootstrap class loader, which is used to load the JVM classes.
> 
> In addition, we decided not to instrument the JAX-RS APIs, as explained above, so I want to close this PR and we will start over once we decide to address #713 .
> 
> Is that OK?

Yes, it's ok. On wildfly and on tomcat it work success. Earlier I added repos:
https://github.com/kananindzya/example-service-b
https://github.com/kananindzya/example-service-a
https://github.com/kananindzya/example-service-a-wildfly
https://github.com/kananindzya/example-service-b-wildfly
where this problem not reproduced.

**Comment by eyalkoren** [comment]
Then we need to figure out what we missed there...
I assume it uses Jersey in those. Strange...

Anyway, closing this one.
Thanks for the validations!!

**Pull request closed without merging**